File: old_contents/boj_codes/C++17/1012.cpp
// https://www.acmicpc.net/problem/1012
// 20201221

#include <cstdio>
#include <deque>

class Node{
public:
    int x;
    int y;

    Node(){
        x = 0;
        y = 0;
    }

    Node(int x_, int y_){
        x = x_;
        y = y_;
    }


    Node operator + (Node& n){
        x = x + n.x;
        y = y + n.y;

        return Node(x, y);
    }

    Node operator - (Node& n){
        x = x - n.x;
        y = y - n.y;

        return Node(x, y);
    }

    void print(){
        printf("%d %d\n", x, y);
    }

};

class Graph{
private:
    Node node;
    Node dir[4];
    int width, height;
    int** cabs;
    bool** visited;

    std::deque<Node> bfs_list;

public:
    int num_bug;
    Graph(int _width, int _height){
        width = _width;
        height = _height;
        num_bug = 0;

        cabs = new int*[height];
        visited = new bool*[height];
        for (int i = 0; i < height; i++){
            cabs[i] = new int[width]();
            visited[i] = new bool[width]();
        }

        dir[0] = Node(0, 1);
        dir[1] = Node(1, 0);
        dir[2] = Node(0, -1);
        dir[3] = Node(-1, 0);

    }

    ~Graph(){
        for (int i = 0; i < height; i++){
            delete [] cabs[i];
            delete [] visited[i];
        }

        delete [] cabs;
        delete [] visited;
        
    }

    void add_cab(int x, int y){
        cabs[y][x] = 1;
    }

    bool check_node(Node& n){
        if(n.x != width)
            return true;
        
        else if(n.y != height - 1){
            n.x = 0;
            n.y += 1;

            return true;
        }

        else
            return false;

    }

    bool check_boundary(Node n){
        if(n.x >= 0 && n.y >= 0 && n.x < width && n.y < height){
            return true;
        }
        
        else{
            return false;
        }
    }

    void BFS(){
        bfs_list.push_back(Node(0, 0));
        visited[0][0] = true;

        while(!bfs_list.empty()){
            node = bfs_list[0];
            bfs_list.pop_front();

            if(cabs[node.y][node.x] == 0){
                node = node + dir[1];
                
                while(check_node(node) && visited[node.y][node.x])
                    node = node + dir[1];

                if(!check_node(node))
                    break;
                
                bfs_list.push_back(node);
                visited[node.y][node.x] = true;

            }

            else{
                num_bug += 1;
                DFS(node);

                while(check_node(node) && visited[node.y][node.x])
                    node = node + dir[1];

                if(!check_node(node))
                    break;
                
                bfs_list.push_back(node);
                visited[node.y][node.x] = true;

            }

        }
    }

    void DFS(Node n){
        for(int i = 0; i < 4; i++){
            n = n + dir[i];

            if(check_boundary(n)){
                if(cabs[n.y][n.x] == 1 && (!visited[n.y][n.x])){
                    visited[n.y][n.x] = 1;

                    DFS(n);
                }

                else
                    visited[n.y][n.x] = 1;
            }
            n = n - dir[i];
        }

        return;

    }

};


int main(){
    int test_case;
    int width, height, num_cabs;
    int x, y;
    scanf("%d", &test_case);

    while(test_case--){
        scanf("%d %d %d", &width, &height, &num_cabs);
        Graph* graph = new Graph(width, height);

        while(num_cabs--){
            scanf("%d %d", &x, &y);
            graph->add_cab(x, y);
        }

        graph->BFS();
        printf("%d\n", graph->num_bug);

        delete graph;

    }

    return 0;
}

File: old_contents/boj_codes/Python3/1012.py
# https://www.acmicpc.net/problem/1012
# 20200921

import sys

sys.setrecursionlimit(50000)

def add_t(x1, y1, x2, y2):
    return x1 + x2, y1 + y2

class Graph:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.cabs = []
        self.visited = [0] * (self.width * self.height)
        self.bug = 0
        self.directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

        for _ in range(self.height):
            self.cabs.append([0] * self.width)

    def add_cab(self, X , Y):
        self.cabs[Y][X] = 1

    def BFS(self):
        queue = [0]
        self.visited[0] = 1

        while len(queue) != 0:
            node = queue.pop(0)

            Y, X = node // self.width, node % self.width
            if self.cabs[Y][X] == 0:
                node += 1
                while node < self.height * self.width and self.visited[node] != 0:
                    node += 1

                if node >= self.height * self.width:
                    break
                queue.append(node)
                self.visited[node] = 1

            else:
                self.bug += 1
                self.adjacent_DFS(X, Y)

                while node < self.height * self.width and self.visited[node] != 0:
                    node += 1

                if node >= self.height * self.width:
                    break
                queue.append(node)
                self.visited[node] = 1

    def adjacent_DFS(self, X, Y):
        for dir1, dir2 in self.directions:
            X_, Y_ = add_t(dir1, dir2, X, Y)
            Z_ = Y_ * self.width + X_

            if X_ >= 0 and X_ < self.width and Y_ >= 0 and Y_ < self.height:
                if self.cabs[Y_][X_] == 1 and \
                        self.visited[Z_] == 0:
                    self.visited[Z_] = 1
                    self.adjacent_DFS(X_, Y_)
                else:
                    self.visited[Z_] = 1


num_test = int(sys.stdin.readline().rstrip())

for i in range(num_test):
    width, height, num_cabs = map(int, sys.stdin.readline().split())
    cab_graph = Graph(width, height)

    for _ in range(num_cabs):
        X, Y = map(int, sys.stdin.readline().split())
        cab_graph.add_cab(X, Y)

    cab_graph.BFS()

    print(cab_graph.bug)
    del cab_graph


